File: src/Aardvark.Base.Rendering/IndexedGeometry.cs
using System;

namespace Aardvark.Base
{
    public enum IndexedGeometryMode
    {
        PointList = 0,
        LineStrip = 1,
        LineList = 2,
        TriangleStrip = 3,
        TriangleList = 4,
    }

    public class IndexedGeometry
    {
        public IndexedGeometryMode Mode;
        public Array IndexArray;
        public SymbolDict<Array> IndexedAttributes;
        public SymbolDict<object> SingleAttributes;

        #region Constructors

        public IndexedGeometry()
            : this(IndexedGeometryMode.TriangleList, null, null, null)
        { }

        public IndexedGeometry(
                Array indexArray,
                SymbolDict<Array> indexedAttributes,
                SymbolDict<object> singleAttributes)
            : this(IndexedGeometryMode.TriangleList, indexArray,
                    indexedAttributes, singleAttributes)
        { }

        public IndexedGeometry(
                IndexedGeometryMode mode,
                Array indexArray,
                SymbolDict<Array> indexedAttributes,
                SymbolDict<object> singleAttributes)
        {
            Mode = mode;
            IndexArray = indexArray;
            IndexedAttributes = indexedAttributes;
            SingleAttributes = singleAttributes;
        }

        #endregion

        #region Properties

        public bool IsIndexed { get { return IndexArray != null; } }

        #endregion
    }
}

File: src/Aardvark.Base.Rendering/IndexedGeometry.fs
namespace Aardvark.Base

open System

type IndexedGeometryMode =
    | PointList = 0
    | LineStrip = 1
    | LineList = 2
    | TriangleStrip = 3
    | TriangleList = 4

type IndexedGeometry =
    class
        val mutable public Mode : IndexedGeometryMode
        val mutable public IndexArray : Array
        val mutable public IndexedAttributes : SymbolDict<Array>
        val mutable public SingleAttributes : SymbolDict<obj>

        member x.IsIndexed = x.IndexArray <> null

        new() = 
            { Mode = IndexedGeometryMode.TriangleList; 
              IndexArray = null; 
              IndexedAttributes = null; 
              SingleAttributes = null }

        new(indexArray, indexedAttributes, singleAttributes) = 
            { Mode = IndexedGeometryMode.TriangleList; 
              IndexArray = indexArray; 
              IndexedAttributes = indexedAttributes; 
              SingleAttributes = singleAttributes }

        new(mode, indexArray, indexedAttributes, singleAttributes) = 
            { Mode = mode
              IndexArray = indexArray; 
              IndexedAttributes = indexedAttributes; 
              SingleAttributes = singleAttributes }
    end
